File: tests/test_phi.py
#!/usr/bin/env -S uv run
"""
Test: test_phi
Tests LLVM PHI node operations

Python equivalent of tests/test_phi.cpp
Must produce identical output to the C++ version.
"""

import llvm


def main():
    with llvm.create_context() as ctx:
        with ctx.create_module("test_phi") as mod:
            i1 = ctx.types.i1
            i32 = ctx.types.i32

            with ctx.create_builder() as builder:
                # ==========================================
                # Function: simple diamond pattern with PHI
                # i32 diamond(i1 cond, i32 a, i32 b)
                # ==========================================
                diamond_ty = ctx.types.function(i32, [i1, i32, i32])
                diamond_func = mod.add_function("diamond", diamond_ty)

                cond = diamond_func.get_param(0)
                a = diamond_func.get_param(1)
                b = diamond_func.get_param(2)
                cond.name = "cond"
                a.name = "a"
                b.name = "b"

                entry = diamond_func.append_basic_block("entry", ctx)
                if_true = diamond_func.append_basic_block("if_true", ctx)
                if_false = diamond_func.append_basic_block("if_false", ctx)
                merge = diamond_func.append_basic_block("merge", ctx)

                # Entry: conditional branch
                builder.position_at_end(entry)
                builder.cond_br(cond, if_true, if_false)

                # True branch: compute a * 2
                builder.position_at_end(if_true)
                a_doubled = builder.mul(a, i32.constant(2), "a_doubled")
                builder.br(merge)

                # False branch: compute b + 1
                builder.position_at_end(if_false)
                b_inc = builder.add(b, i32.constant(1), "b_inc")
                builder.br(merge)

                # Merge: PHI node to select result
                builder.position_at_end(merge)
                phi = builder.phi(i32, "result")

                # Add incoming values
                phi.add_incoming(a_doubled, if_true)
                phi.add_incoming(b_inc, if_false)

                builder.ret(phi)

                # ==========================================
                # Function: loop with PHI (sum 1 to n)
                # i32 sum_to_n(i32 n)
                # ==========================================
                sum_ty = ctx.types.function(i32, [i32])
                sum_func = mod.add_function("sum_to_n", sum_ty)

                n = sum_func.get_param(0)
                n.name = "n"

                sum_entry = sum_func.append_basic_block("entry", ctx)
                loop = sum_func.append_basic_block("loop", ctx)
                exit_bb = sum_func.append_basic_block("exit", ctx)

                # Entry: branch to loop
                builder.position_at_end(sum_entry)
                builder.br(loop)

                # Loop header with PHIs
                builder.position_at_end(loop)
                i_phi = builder.phi(i32, "i")
                sum_phi = builder.phi(i32, "sum")

                # Loop body: sum += i; i++
                new_sum = builder.add(sum_phi, i_phi, "new_sum")
                new_i = builder.add(i_phi, i32.constant(1), "new_i")

                # Loop condition: i <= n
                loop_cond = builder.icmp(llvm.IntPredicate.SLE, new_i, n, "loop_cond")
                builder.cond_br(loop_cond, loop, exit_bb)

                # Add incoming values to PHIs
                # From entry: i=1, sum=0
                i_phi.add_incoming(i32.constant(1), sum_entry)
                i_phi.add_incoming(new_i, loop)

                sum_phi.add_incoming(i32.constant(0), sum_entry)
                sum_phi.add_incoming(new_sum, loop)

                # Exit: return sum
                builder.position_at_end(exit_bb)
                builder.ret(new_sum)

            # Verify module
            if not mod.verify():
                print(f"; Verification failed: {mod.get_verification_error()}")
                return 1

            # Print diagnostic comments
            print("; Test: test_phi")
            print(";")
            print("; Diamond pattern PHI:")
            print(f";   phi incoming count: {phi.num_incoming}")

            # Get incoming values and blocks
            for i in range(phi.num_incoming):
                val = phi.get_incoming_value(i)
                blk = phi.get_incoming_block(i)
                print(f";   incoming[{i}]: value={val.name}, block={blk.name}")

            print(";")
            print("; Loop PHIs:")
            print(f";   i_phi incoming count: {i_phi.num_incoming}")
            print(f";   sum_phi incoming count: {sum_phi.num_incoming}")

            print()

            # Print module IR
            print(mod.to_string(), end="")

    return 0


if __name__ == "__main__":
    exit(main())

File: tests/test_phi.cpp
/**
 * Test: test_phi
 * Tests LLVM PHI node operations
 *
 * LLVM-C APIs covered:
 * - LLVMBuildPhi()
 * - LLVMAddIncoming()
 * - LLVMCountIncoming(), LLVMGetIncomingValue(), LLVMGetIncomingBlock()
 */

#include <cstdio>
#include <llvm-c/Analysis.h>
#include <llvm-c/Core.h>

int main() {
  LLVMContextRef ctx = LLVMContextCreate();
  LLVMModuleRef mod = LLVMModuleCreateWithNameInContext("test_phi", ctx);

  LLVMTypeRef i1 = LLVMInt1TypeInContext(ctx);
  LLVMTypeRef i32 = LLVMInt32TypeInContext(ctx);

  LLVMBuilderRef builder = LLVMCreateBuilderInContext(ctx);

  // ==========================================
  // Function: simple diamond pattern with PHI
  // i32 diamond(i1 cond, i32 a, i32 b)
  // ==========================================
  LLVMTypeRef diamond_params[] = {i1, i32, i32};
  LLVMTypeRef diamond_ty = LLVMFunctionType(i32, diamond_params, 3, 0);
  LLVMValueRef diamond_func = LLVMAddFunction(mod, "diamond", diamond_ty);

  LLVMValueRef cond = LLVMGetParam(diamond_func, 0);
  LLVMValueRef a = LLVMGetParam(diamond_func, 1);
  LLVMValueRef b = LLVMGetParam(diamond_func, 2);
  LLVMSetValueName2(cond, "cond", 4);
  LLVMSetValueName2(a, "a", 1);
  LLVMSetValueName2(b, "b", 1);

  LLVMBasicBlockRef entry =
      LLVMAppendBasicBlockInContext(ctx, diamond_func, "entry");
  LLVMBasicBlockRef if_true =
      LLVMAppendBasicBlockInContext(ctx, diamond_func, "if_true");
  LLVMBasicBlockRef if_false =
      LLVMAppendBasicBlockInContext(ctx, diamond_func, "if_false");
  LLVMBasicBlockRef merge =
      LLVMAppendBasicBlockInContext(ctx, diamond_func, "merge");

  // Entry: conditional branch
  LLVMPositionBuilderAtEnd(builder, entry);
  LLVMBuildCondBr(builder, cond, if_true, if_false);

  // True branch: compute a * 2
  LLVMPositionBuilderAtEnd(builder, if_true);
  LLVMValueRef a_doubled =
      LLVMBuildMul(builder, a, LLVMConstInt(i32, 2, 0), "a_doubled");
  LLVMBuildBr(builder, merge);

  // False branch: compute b + 1
  LLVMPositionBuilderAtEnd(builder, if_false);
  LLVMValueRef b_inc =
      LLVMBuildAdd(builder, b, LLVMConstInt(i32, 1, 0), "b_inc");
  LLVMBuildBr(builder, merge);

  // Merge: PHI node to select result
  LLVMPositionBuilderAtEnd(builder, merge);
  LLVMValueRef phi = LLVMBuildPhi(builder, i32, "result");

  // Add incoming values
  LLVMValueRef incoming_vals[] = {a_doubled, b_inc};
  LLVMBasicBlockRef incoming_blocks[] = {if_true, if_false};
  LLVMAddIncoming(phi, incoming_vals, incoming_blocks, 2);

  LLVMBuildRet(builder, phi);

  // ==========================================
  // Function: loop with PHI (sum 1 to n)
  // i32 sum_to_n(i32 n)
  // ==========================================
  LLVMTypeRef sum_params[] = {i32};
  LLVMTypeRef sum_ty = LLVMFunctionType(i32, sum_params, 1, 0);
  LLVMValueRef sum_func = LLVMAddFunction(mod, "sum_to_n", sum_ty);

  LLVMValueRef n = LLVMGetParam(sum_func, 0);
  LLVMSetValueName2(n, "n", 1);

  LLVMBasicBlockRef sum_entry =
      LLVMAppendBasicBlockInContext(ctx, sum_func, "entry");
  LLVMBasicBlockRef loop = LLVMAppendBasicBlockInContext(ctx, sum_func, "loop");
  LLVMBasicBlockRef exit_bb =
      LLVMAppendBasicBlockInContext(ctx, sum_func, "exit");

  // Entry: branch to loop
  LLVMPositionBuilderAtEnd(builder, sum_entry);
  LLVMBuildBr(builder, loop);

  // Loop header with PHIs
  LLVMPositionBuilderAtEnd(builder, loop);
  LLVMValueRef i_phi = LLVMBuildPhi(builder, i32, "i");
  LLVMValueRef sum_phi = LLVMBuildPhi(builder, i32, "sum");

  // Loop body: sum += i; i++
  LLVMValueRef new_sum = LLVMBuildAdd(builder, sum_phi, i_phi, "new_sum");
  LLVMValueRef new_i =
      LLVMBuildAdd(builder, i_phi, LLVMConstInt(i32, 1, 0), "new_i");

  // Loop condition: i <= n
  LLVMValueRef loop_cond =
      LLVMBuildICmp(builder, LLVMIntSLE, new_i, n, "loop_cond");
  LLVMBuildCondBr(builder, loop_cond, loop, exit_bb);

  // Add incoming values to PHIs
  // From entry: i=1, sum=0
  LLVMValueRef i_incoming_vals[] = {LLVMConstInt(i32, 1, 0), new_i};
  LLVMBasicBlockRef i_incoming_blocks[] = {sum_entry, loop};
  LLVMAddIncoming(i_phi, i_incoming_vals, i_incoming_blocks, 2);

  LLVMValueRef sum_incoming_vals[] = {LLVMConstInt(i32, 0, 0), new_sum};
  LLVMBasicBlockRef sum_incoming_blocks[] = {sum_entry, loop};
  LLVMAddIncoming(sum_phi, sum_incoming_vals, sum_incoming_blocks, 2);

  // Exit: return sum
  LLVMPositionBuilderAtEnd(builder, exit_bb);
  LLVMBuildRet(builder, new_sum);

  LLVMDisposeBuilder(builder);

  // Verify module
  char *error = nullptr;
  if (LLVMVerifyModule(mod, LLVMReturnStatusAction, &error)) {
    fprintf(stderr, "; Verification failed: %s\n", error);
    LLVMDisposeMessage(error);
    LLVMDisposeModule(mod);
    LLVMContextDispose(ctx);
    return 1;
  }
  LLVMDisposeMessage(error);

  // Print diagnostic comments
  printf("; Test: test_phi\n");
  printf(";\n");
  printf("; Diamond pattern PHI:\n");
  printf(";   phi incoming count: %u\n", LLVMCountIncoming(phi));

  // Get incoming values and blocks
  for (unsigned i = 0; i < LLVMCountIncoming(phi); i++) {
    LLVMValueRef val = LLVMGetIncomingValue(phi, i);
    LLVMBasicBlockRef blk = LLVMGetIncomingBlock(phi, i);
    size_t len;
    const char *val_name = LLVMGetValueName2(val, &len);
    const char *blk_name = LLVMGetBasicBlockName(blk);
    printf(";   incoming[%u]: value=%s, block=%s\n", i, val_name, blk_name);
  }

  printf(";\n");
  printf("; Loop PHIs:\n");
  printf(";   i_phi incoming count: %u\n", LLVMCountIncoming(i_phi));
  printf(";   sum_phi incoming count: %u\n", LLVMCountIncoming(sum_phi));

  printf("\n");

  // Print module IR
  char *ir = LLVMPrintModuleToString(mod);
  printf("%s", ir);
  LLVMDisposeMessage(ir);

  // Cleanup
  LLVMDisposeModule(mod);
  LLVMContextDispose(ctx);

  return 0;
}
